src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs

```diff
@@ -1260,6 +1321,66 @@ private SourceColumnMetadata GetColumnMetadata(int ordinal)
                     method = ValueMethod.GetValue;
                 }
             }
+            else if (_rowSourceType == ValueSourceType.IDataReader)
+            {
+                isSqlType = false;
+                isDataFeed = false;
+
+                Type t = ((IDataReader)_rowSource).GetFieldType(ordinal);
+
+                if (t == typeof(bool))
```

Comment:
@cheenamalhotra @Wraith2 

Thinking about this in the context of the `IsDbNull` check above -- I think we need to add an OR check to the corresponding nullable type (e.g. `t == typeof(bool) || t == typeof(bool?`)

I think we have two routes we can go here.

1. We add the OR check and invoke `IsDbNull` on each value.  Pros: prevents nullable types from being boxed.  Cons: have to invoke IsDbNull (per impact TBD) and requires data reader implementations to correct determine result of `IsDbNull`
2. Leave this as is and we don't have to invoke `IsDbNull` on each value.  Nullable value types would get boxed.

I would prefer option 1 if possible.

I will measure the option 1 approach as discussed and we will see how it shakes out.

Reply:
I don't believe that GetFieldType can return a Nullable<bool>, at least it can't in the SqlClient implementation. We don't conflate sql and language nulls (rightly imo) so I don't think you need to change this. If a non-SqlClient data source provides a Nullable<bool> then I think boxing is probably appropriate.

Reply:
A nullable int column should have GetFieldType of int, nullability is not part of the type in sql.